### ISSUE TITLE
Bug with autograd for some linear algebra operations

### ISSUE THREAD

**Issue by adofarsi** [pytorch/pytorch#133606]
### 🐛 Describe the bug

I have an ML model that involves certain linear algebra operations and back propagation has failed for my model. After some investigations it seems that autograd fails to differentiate a simple linear algebra operation and I don't know what to do. Here is a minimal failing example:
```python3
def f(x):
    # Compute trace
    mask = torch.eye(2)
    I = (x * mask[None, ...]).sum(axis=(1, 2))
    # Compute determinant
    J = torch.linalg.det(x)
    # Compute f
    return I + J**0.5

torch.autograd.gradcheck(f, torch.rand(9, 2, 2, dtype=torch.double, requires_grad=True))
```
I'm basically trying to compute the sum of the trace and the square root of the determinant of a matrix.
Differentiation seems to work if I make some simple adjustments to the above function, such as considering a 2x2 matrix as input and using `torch.trace()` to calculate the trace.  Does anyone have some ideas on how to solve this problem? Thanks!

### Versions

````
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.16 (default, Dec  7 2022, 01:12:13)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.27
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               142
Model name:          Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
Stepping:            10
CPU MHz:             2483.549
CPU max MHz:         4200.0000
CPU min MHz:         400.0000
BogoMIPS:            4199.88
Virtualisation:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            8192K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] numpy==1.24.0
[pip3] torch==2.3.0+cpu
[pip3] torch_geometric==2.5.3
[conda] blas                      1.0                         mkl  
[conda] cpuonly                   1.0                           0    pytorch
[conda] cudatoolkit               8.0                           3    https://repo.anaconda.com/pkgs/free
[conda] mkl                       2019.3                      199  
[conda] mkl-service               1.1.2            py37he904b0f_5  
[conda] mkl_fft                   1.0.10           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] numpy                     1.16.2           py37h7e9f1db_0  
[conda] numpy-base                1.16.2           py37hde5b4d6_0  
[conda] numpydoc                  0.8.0                    py37_0  
[conda] pytorch                   1.7.1               py3.7_cpu_0  [cpuonly]  pytorch
[conda] torchaudio                0.7.2                      py37    pytorch
[conda] torchvision               0.8.2                  py37_cpu  [cpuonly]  pytorch
[conda] torchvision-cpu           0.2.2                      py_3    pytorch
```

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @jianyuh @mruberry @walterddr @xwang233 @Lezcano

**Comment by soulitzer** [comment]
This is expected behavior because you are trying to take the sqrt of the determinant which is not always positive. You should observe that even with 2x2 matrix, your function should only succeed about half the time depending on the seed. The reason 9 x 2 x2 is almost guaranteed to fail is because as you increase your batch size, the gradcheck would only succeed if ALL determinants are greater than zero.
